### PR TITLE
Desktop Username password

### PR DESCRIPTION
**Pull request by jigar-f**

### CODE REVIEW

**Comment by jigar-f** [comment]
@atavism You start reviewing this PR.

File: internalsdk/auth/srp.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	if err != nil {
+		return nil, err
+	}
+	signUpRequestBody := &protos.SignupRequest{
+		Email:                 lowerCaseEmail,
+		Salt:                  salt,
+		Verifier:              verifierKey.Bytes(),
+		SkipEmailConfirmation: true,
+	}
+	log.Debugf("Sign up request email %v, salt %v verifier %v verifiter in bytes %v", lowerCaseEmail, salt, verifierKey, verifierKey.Bytes())
```
Comment (atavism) [inline internalsdk/auth/srp.go:120]: I think once we have confirmation this is working well, we should avoid logging the salt, verifier, etc. here in case the user submits their logs to us

Comment (jigar-f) [reply internalsdk/auth/srp.go:120]: correct, I am planning to remove logs.

File: desktop/lib.go [hunk 2]
```diff
@@ -98,6 +103,13 @@ func start() {
 		}
 	}()
 
+	go func() {
+		err = fetchUserData()
+		if err != nil {
+			log.Errorf("error while fetching user data: %v", err)
+		}
+	}()
```
Comment (atavism) [inline desktop/lib.go:111]: I would maybe just simplify this to..

```go
go fetchUserData()
```

since we are already logging the error in `fetchUserData`

File: desktop/lib.go [hunk 3]
```diff
@@ -746,4 +890,322 @@ func handleSignals(a *app.App) {
...
+		return sendError(err)
+	}
+	prepareRequestBody := &protos.CompleteRecoveryByEmailRequest{
+		Email:       lowerCaseEmail,
+		Code:        C.GoString(code),
+		NewSalt:     newsalt,
+		NewVerifier: verifierKey.Bytes(),
+	}
+
+	log.Debugf("new Verifier %v and salt %v", verifierKey.Bytes(), newsalt)
```
Comment (atavism) [inline desktop/lib.go:1084]: Avoid logging the verifier and salt here as well

File: desktop/lib.go [hunk 4]
```diff
@@ -746,4 +890,322 @@ func handleSignals(a *app.App) {
...
+		LegacyToken:  token,
+		LegacyUserID: userId,
+	}
+	log.Debugf("Sign out request %+v", signoutData)
+	loggedOut, logoutErr := authClient.SignOut(context.Background(), signoutData)
+	if logoutErr != nil {
+		return sendError(log.Errorf("Error while signing out %v", logoutErr))
+	}
+	if !loggedOut {
+		return sendError(log.Errorf("Error while signing out %v", logoutErr))
```
Comment (atavism) [inline desktop/lib.go:1009]: I think logoutErr is nil here

**Comment by atavism** [comment]
Screenshot 2024-07-09 at 6 07 30 AM

I think we may need to increase the height on the input fields on the sign in pages (cc @Derekf5)

**Comment by atavism** [comment]
I keep running into this issue switching tabs/navigating to different pages that causes the app to crash..  I can dig into it, but just wondering if you encountered anything like this?

```
signal 16 received but handler not on signal stack
mp.gsignal stack [0x14000084000 0x1400008c000], mp.g0 stack [0x16d374000 0x16d577000], sp=0x14004d87848
fatal error: non-Go code set up signal handler without SA_ONSTACK flag

runtime stack:
runtime.throw({0x15dd3980d?, 0x0?})
	runtime/panic.go:1023 +0x40 fp=0x14004d877a0 sp=0x14004d87770 pc=0x15d013100
runtime.sigNotOnStack(0x10, 0x14004d87848, 0x14000080008)
	runtime/signal_unix.go:1065 +0x118 fp=0x14004d877d0 sp=0x14004d877a0 pc=0x15d02db68
runtime.adjustSignalStack(0x10, 0x14000080008, 0x14004d87878)
	runtime/signal_unix.go:592 +0x25c fp=0x14004d87840 sp=0x14004d877d0 pc=0x15d02c9cc
runtime.sigtrampgo(0x10, 0x14004d879e0, 0x14004d87a48)
	runtime/signal_unix.go:480 +0x8c fp=0x14004d878c0 sp=0x14004d87840 pc=0x15d02c50c
runtime.sigtrampgo(0x10, 0x14004d879e0, 0x14004d87a48)
	<autogenerated>:1 +0x1c fp=0x14004d878f0 sp=0x14004d878c0 pc=0x15d05283c
runtime.sigtramp()
	runtime/sys_darwin_arm64.s:227 +0x4c fp=0x14004d879b0 sp=0x14004d878f0 pc=0x15d05130c

goroutine 34 gp=0x1400010c380 m=7 mp=0x14000108008 [syscall, locked to thread]:
runtime.goexit({})
	runtime/asm_arm64.s:1222 +0x4 fp=0x1400008cfe0 sp=0x1400008cfe0 pc=0x15d050244

goroutine 50 gp=0x140000021c0 m=nil [chan receive]:
runtime.gopark(0x14000461f38?, 0x15db39a9c?, 0x20?, 0x81?, 0x15e2bd0c0?)
	runtime/proc.go:402 +0xc8 fp=0x140001e2ec0 sp=0x140001e2ea0 pc=0x15d015f58
runtime.chanrecv(0x140000482a0, 0x14000461fa8, 0x1)
	runtime/chan.go:583 +0x404 fp=0x140001e2f40 sp=0x140001e2ec0 pc=0x15cfdfa84
runtime.chanrecv2(0x15e09ed60?, 0x14000808120?)
	runtime/chan.go:447 +0x14 fp=0x140001e2f70 sp=0x140001e2f40 pc=0x15cfdf664
github.com/getlantern/flashlight/v7/config.InitWithURLs.func2()
	github.com/getlantern/flashlight/v7@v7.6.87/config/initializer.go:81 +0x6c fp=0x140001e2fd0 sp=0x140001e2f70 pc=0x15db159bc
runtime.goexit({})
	runtime/asm_arm64.s:1222 +0x4 fp=0x140001e2fd0 sp=0x140001e2fd0 pc=0x15d050244
created by github.com/getlantern/flashlight/v7/config.InitWithURLs in goroutine 76
	github.com/getlantern/flashlight/v7@v7.6.87/config/initializer.go:80 +0x17c
```

**Comment by jigar-f** [comment]
> I keep running into this issue switching tabs/navigating to different pages that causes the app to crash.. I can dig into it, but just wondering if you encountered anything like this?
> 
> ```
> signal 16 received but handler not on signal stack
> mp.gsignal stack [0x14000084000 0x1400008c000], mp.g0 stack [0x16d374000 0x16d577000], sp=0x14004d87848
> fatal error: non-Go code set up signal handler without SA_ONSTACK flag
> 
> runtime stack:
> runtime.throw({0x15dd3980d?, 0x0?})
> 	runtime/panic.go:1023 +0x40 fp=0x14004d877a0 sp=0x14004d87770 pc=0x15d013100
> runtime.sigNotOnStack(0x10, 0x14004d87848, 0x14000080008)
> 	runtime/signal_unix.go:1065 +0x118 fp=0x14004d877d0 sp=0x14004d877a0 pc=0x15d02db68
> runtime.adjustSignalStack(0x10, 0x14000080008, 0x14004d87878)
> 	runtime/signal_unix.go:592 +0x25c fp=0x14004d87840 sp=0x14004d877d0 pc=0x15d02c9cc
> runtime.sigtrampgo(0x10, 0x14004d879e0, 0x14004d87a48)
> 	runtime/signal_unix.go:480 +0x8c fp=0x14004d878c0 sp=0x14004d87840 pc=0x15d02c50c
> runtime.sigtrampgo(0x10, 0x14004d879e0, 0x14004d87a48)
> 	<autogenerated>:1 +0x1c fp=0x14004d878f0 sp=0x14004d878c0 pc=0x15d05283c
> runtime.sigtramp()
> 	runtime/sys_darwin_arm64.s:227 +0x4c fp=0x14004d879b0 sp=0x14004d878f0 pc=0x15d05130c
> 
> goroutine 34 gp=0x1400010c380 m=7 mp=0x14000108008 [syscall, locked to thread]:
> runtime.goexit({})
> 	runtime/asm_arm64.s:1222 +0x4 fp=0x1400008cfe0 sp=0x1400008cfe0 pc=0x15d050244
> 
> goroutine 50 gp=0x140000021c0 m=nil [chan receive]:
> runtime.gopark(0x14000461f38?, 0x15db39a9c?, 0x20?, 0x81?, 0x15e2bd0c0?)
> 	runtime/proc.go:402 +0xc8 fp=0x140001e2ec0 sp=0x140001e2ea0 pc=0x15d015f58
> runtime.chanrecv(0x140000482a0, 0x14000461fa8, 0x1)
> 	runtime/chan.go:583 +0x404 fp=0x140001e2f40 sp=0x140001e2ec0 pc=0x15cfdfa84
> runtime.chanrecv2(0x15e09ed60?, 0x14000808120?)
> 	runtime/chan.go:447 +0x14 fp=0x140001e2f70 sp=0x140001e2f40 pc=0x15cfdf664
> github.com/getlantern/flashlight/v7/config.InitWithURLs.func2()
> 	github.com/getlantern/flashlight/v7@v7.6.87/config/initializer.go:81 +0x6c fp=0x140001e2fd0 sp=0x140001e2f70 pc=0x15db159bc
> runtime.goexit({})
> 	runtime/asm_arm64.s:1222 +0x4 fp=0x140001e2fd0 sp=0x140001e2fd0 pc=0x15d050244
> created by github.com/getlantern/flashlight/v7/config.InitWithURLs in goroutine 76
> 	github.com/getlantern/flashlight/v7@v7.6.87/config/initializer.go:80 +0x17c
> ```

@atavism  I am getting this issue multiple times at random places in app, I am not sure whats's going on here, If you take look, That would be great, There is something related to go signals.

File: desktop/lib.go (hunk 4 above)
Comment (jigar-f) [reply desktop/lib.go:1009]: Done.

File: desktop/lib.go (hunk 2 above)
Comment (jigar-f) [reply desktop/lib.go:111]: Done.

**Review by atavism — APPROVED** [review]

**Comment by jigar-f** [comment]
@atavism thanks for approving, Did you able to find out about the crash?

**Comment by atavism** [comment]
@jigar-f I haven't been able to figure out what's causing the app to crash. I actually haven't experienced the same issue in awhile testing the latest changes. So I think we can just wait to see if it resurfaces

**Comment by atavism** [comment]
I discovered an issue where it looks like the timer in the VPNChangeNotifier isn't being stopped when it reaches the max number of ticks..

```
Jul 16 16:34:43.809 - 0m20s DEBUG app.settings: settings.go:670 Could not get value for userPro
flutter: startup status proxy true config true success false
flutter: flashlight fail initialized
Jul 16 16:34:44.810 - 0m21s DEBUG app.settings: settings.go:670 Could not get value for userPro
flutter: startup status proxy true config true success false
flutter: flashlight fail initialized
Jul 16 16:34:45.809 - 0m22s DEBUG app.settings: settings.go:670 Could not get value for userPro
flutter: startup status proxy true config true success false
flutter: flashlight fail initialized
Jul 16 16:34:46.810 - 0m23s DEBUG app.settings: settings.go:670 Could not get value for userPro
flutter: startup status proxy true config true success false
flutter: flashlight fail initialized
Jul 16 16:34:47.810 - 0m24s DEBUG app.settings: settings.go:670 Could not get value for userPro
flutter: startup status proxy true config true success false
flutter: flashlight fail initialized
Jul 16 16:34:48.810 - 0m25s DEBUG app.settings: settings.go:670 Could not get value for userPro
flutter: startup status proxy true config true success false
flutter: flashlight fail initialized
```

**Comment by jigar-f** [comment]
> t

This is not an issue, We run the timer till the flashlight is initialized.

**Comment by jigar-f** [comment]
```
signal 16 received but handler not on signal stack
mp.gsignal stack [0x14000084000 0x1400008c000], mp.g0 stack [0x16ba88000 0x16bc8b000], sp=0x14000857678
fatal error: non-Go code set up signal handler without SA_ONSTACK flag
```
@atavism I am still getting crash, Any idea whats going on?

**Comment by atavism** [comment]
> @atavism I am still getting crash, Any idea whats going on?

No, I am still having trouble getting the app to crash again myself. What are you doing just before the crash happens -- switching tabs, turning the VPN on, etc? It might be helpful to add additional logging to try to pinpoint where the crash is happening

**Comment by jigar-f** [comment]
It crashes for me at random places, I have already added some logging to track it down.

**Comment by jigar-f** [comment]
@atavism [This](https://pkg.go.dev/os/signal#hdr-Types_of_signals) might be the issue

File: desktop/lib.go [hunk 5]
```diff
@@ -80,15 +81,17 @@ func start() {
 
 	cdir := configDir(&flags)
 	settings := loadSettings(cdir)
-	proClient = proclient.NewClient(fmt.Sprintf("https://%s", common.ProAPIHost), &webclient.Opts{
+	webclientOpts := &webclient.Opts{
 		HttpClient: &http.Client{
-			Transport: proxied.ParallelForIdempotent(),
+			Transport: proxied.Fronted(30 * time.Second),
```
Comment (atavism) [inline desktop/lib.go:86]: Why the change to use a proxied.Fronted http.RoundTripper here..? We want to continue using `proxied. ParallelForIdempotent`

File: desktop/lib.go [hunk 6]
```diff
@@ -122,7 +127,7 @@ func start() {
 	}()
 
 	golog.SetPrepender(logging.Timestamped)
-	handleSignals(a)
+	// handleSignals(a)
```
Comment (atavism) [inline desktop/lib.go:130]: I don't think this code to handle signals is the problem area since it's been around and hasn't caused the app to crash previously. It's likely some code that changed recently..

For instance, I wonder if we maybe introduced a bug somewhere fetching and storing user data.. the fact we see this error consistently in the logs suggests an issue where a thread may be attempting to relock a mutex (like [this one](https://github.com/getlantern/lantern-client/blob/main/desktop/app/pro.go#L19) without first unlocking it:

```
goroutine 283 gp=0x14000ba3a40 m=nil [sync.Mutex.Lock]:
```

have you tried commenting out any of the new code or calls that are being made to see if the app still crashes with those disabled?

Comment (atavism) [reply desktop/lib.go:130]: Have we checked if the app crashes on main? Let me do that now..

Comment (atavism) [reply desktop/lib.go:130]: Ok, this is interesting: the app is crashing for me on main, but it no longer does if I go back to commit 042186e7d2510b7185ac3398b5284535d4cb0274. So I'm pretty sure the crash is unrelated to the changes here.

Given the crash is unrelated, let's revert the changes to disable signal handling, continue using `ParallelForIdempotent`, and merge this? Then we can debug the issue further on main.

File: desktop/lib.go (hunk 5 above)
Comment (jigar-f) [reply desktop/lib.go:86]: Yep, But there was an issue with `ParallelForIdempotent` this make two requests for each call, and the first call fails every time, even with the proxy on or off, I might need to do more research on this, Will next week.

File: desktop/lib.go (hunk 6 above)
Comment (jigar-f) [reply desktop/lib.go:130]: The error you are seeing is different then mine, this looks like some nil pointer issue, and mine is something with signal, you can check slack for more info.

Comment (jigar-f) [reply desktop/lib.go]: Sure, I have re-enbale signal and added proxied.ParallelForIdempotent

Comment (jigar-f) [reply desktop/lib.go]: I am fine with merging.

Comment (jigar-f) [reply desktop/lib.go]: To me it looks like this might be crash, Since this actually throws painc.
```
runtime stack:
runtime.throw({0x15c9fc20e?, 0x0?})
	runtime/panic.go:1023 +0x40 fp=0x14000051490 sp=0x14000051460 pc=0x15bce13a0
runtime.sigNotOnStack(0x10, 0x14000051538, 0x14000080008)
	runtime/signal_unix.go:1065 +0x118 fp=0x140000514c0 sp=0x14000051490 pc=0x15bcfbe08
runtime.adjustSignalStack(0x10, 0x14000080008, 0x14000051568)
	runtime/signal_unix.go:592 +0x25c fp=0x14000051530 sp=0x140000514c0 pc=0x15bcfac6c
runtime.sigtrampgo(0x10, 0x140000516d0, 0x14000051738)
	runtime/signal_unix.go:480 +0x8c fp=0x140000515b0 sp=0x14000051530 pc=0x15bcfa7ac
runtime.sigtrampgo(0x10, 0x140000516d0, 0x14000051738)
	<autogenerated>:1 +0x1c fp=0x140000515e0 sp=0x140000515b0 pc=0x15bd20adc
runtime.sigtramp()
	runtime/sys_darwin_arm64.s:227 +0x4c fp=0x140000516a0 sp=0x140000515e0 pc=0x15bd1f5ac
```

Comment (jigar-f) [reply desktop/lib.go]: Here is ticket-:https://github.com/getlantern/engineering/issues/1504

**Comment by jigar-f** [comment]
@atavism All changes are added, SHould I go ahead and merge this?

File: desktop/lib.go (hunk 6 above)
Comment (atavism) [reply desktop/lib.go]: Thanks, @jigar-f! I will take a closer look at this over the weekend

**Comment by atavism** [comment]
> @atavism All changes are added, SHould I go ahead and merge this?

Thanks for making those changes! Sounds great! Merging..
